src/compiler/checker.ts

```diff
@@ -23001,6 +23003,15 @@ namespace ts {
             }
         }
 
+        function getContextualTypeForPropertyDeclaration(declaration: PropertyDeclaration): Type | undefined {
+            const parent = declaration.parent.parent;
+            const parentType = isVariableDeclaration(parent) && getContextualTypeForVariableLikeDeclaration(parent);
+            if (!parentType) return undefined;
+            const resolved = getPropertyOfType(parentType, getSymbolOfNode(declaration).escapedName);

```

Comment:
Pretty sure this needs a check for the `PropertyDeclaration` in question being `static` somewhere before doing this lookup, no? Also, I don't think `getPropertyOfType(parentType, getSymbolOfNode(declaration).escapedName)` is going to handle late-bound names like
```
interface I {
    x: { a: "a" };
}
const x = "x";
let c: I = class {
    // should typecheck the same as the last line
    static [x] = { a: "a" };
};
c[x] = { a: "a" };

```
well - you may need to use a different pattern that relies on `getIndexedAccessType` instead.

Reply:
Hm, looks like we largely don't handle late bound names much in contextual typing (we do in like 1 place, maybe, but it's irregular), but we _do_ have a `getTypeOfPropertyOfContextualType` helper that does a bunch of extra work for unions that you should use instead.

Reply:
1. I planned to test for `static`, but eventually forgot to add it...  Done now.
2. Re dynamic names -- I tried it, and it worked -- but I still switched the code to use `getTypeOfPropertyOfContextualType`.